lib/spack/spack/traverse.py

```diff
@@ -96,18 +96,80 @@ def __init__(self, visitor, key=id, visited=None):
         self.visited = set() if visited is None else visited
         self.key = key
 
-    def accept(self, node):
-        return self.visitor.accept(node)
+    def accept(self, item):
+        return self.visitor.accept(item)
 
-    def neighbors(self, node):
+    def neighbors(self, item):
         # Covering edges means: drop dependencies of visited nodes.
-        key = self.key(node.edge.spec)
+        key = self.key(item.edge.spec)
 
         if key in self.visited:
             return []
 
         self.visited.add(key)
-        return self.visitor.neighbors(node)
+        return self.visitor.neighbors(item)
+
+
+class TopoVisitor(object):
+    """Visitor that can be used in depth-first search to generate
+    a topological ordered list of vertices (we keep track of edges).
+
+    Algorithm based on "Section 22.4: Topological sort", Introduction to Algorithms
+    (2001, 2nd edition) by Cormen, Thomas H.; Leiserson, Charles E.; Rivest, Ronald L.;
+    Stein, Clifford.
+
+    The algorithm in plain English: realize that (1) we only append a vertex A before
```

Comment:
I think the following description would be clearer, but I'm happy to iterate on what's here:

Perform a depth-first search; append a node only after visiting all of its children. This produces a list where the children of a node are always guaranteed to come before a node. This list can be reversed to produce a topological ordering.

Again, this depends on being used in tandem with `traverse_depth_first_with_visitor`, and to that end I think it would be clearer if this class were defined inside of the method body of `traverse_edges_topo`.

Reply:
I don't agree with the latter, it's not an exposed class in the module, but for advanced use cases you might want to override parts of it, like `neighbors`.